### PR TITLE
Add DP API Cache

### LINKED ISSUE #1: Create a DP API Token Cache (Windows Only)
Create a TokenCache class which extends the existing [SerializableTokenCache](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/f76f3c37c6bf7747971947d300f0a50d3ec701f7/msal/token_cache.py#L204) to handle interacting with a token cache which is persisted in a file protected by Windows provided encryption logic.

The goal is to bring this Python library to parity with the .NET functionality found in the [MsalCacheStorage class.](https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/blob/cf9ca48e0295efca0562a743e7758d1c59442f22/src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheStorage.cs#L16)

### PR DESCRIPTION
**Pull request by marstr**
Closes #1

### CODE REVIEW

File: msal_extensions/_windows.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
+        self._lock_location = self._cache_location + '.lockfile'
+
+        entropy = _WindowsTokenCache.DEFAULT_ENTROPY
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy']
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        self._last_sync = 0  # _last_sync is a Unixtime
+
+    def _has_state_changed(self):
```
Comment (rayluo) [inline msal_extensions/_windows.py]: My bad. I used the name `has_state_changed` in `SerializableTokenCache`. At a hindsight, I would probably rename it to `has_memory_state_changed` soon.
And then your function here should really be `has_file_state_changed()`. And these 2 flags are exactly the opposite meaning.

File: msal_extensions/_windows.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        # type: () -> Bool
+        try:
+            return self.has_state_changed or self._last_sync < os.path.getmtime(self._cache_location)
```
Comment (rayluo) [inline msal_extensions/_windows.py]: Now you might want to skip the `self.has_memory_state_changed` check here, because ideally it shouldn't happen in here in this class.

File: msal_extensions/_windows.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise e
+            return False
+
+    def add(self, event, **kwargs):
+        if self._has_state_changed:
```
Comment (rayluo) [inline msal_extensions/_windows.py]: You mean `self._has_state_changed`**()**?

File: msal_extensions/_windows.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).add(event, **kwargs)
+        self._write()
+
+    def update_rt(self, rt_item, new_rt):
```
Comment (rayluo) [inline msal_extensions/_windows.py]: I know you were not aware of this: we've been working on a new PR in MSAL Python which would introduce a bunch of new `remove_xx(...)` methods, but here comes the good news! We [optimized it](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/41/files#diff-6f83eec40714321407e16ebed558f014R287) for you, so that you would only need to wrap one `modify()` instead of all those `update_xx()` and `remove_xx()`. You would need to change these 2 methods here, after we release MSAL Python 0.4.x, soon.

File: msal_extensions/_windows.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        if self.has_state_changed:
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).update_rt(rt_item, new_rt)
+        self._write()
+
+    def remove_rt(self, rt_item):
```
Comment (rayluo) [inline msal_extensions/_windows.py]: Same as above `update_rt()`.

File: msal_extensions/_windows.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        if self._has_state_changed:
+            try:
+                self._read()
+            except OSError as e:
+                if e.errno != errno.ENOENT:
+                    raise e
+        super(_WindowsTokenCache, self).remove_rt(rt_item)
+        self._write()
+
+    def find(self, credential_type, target=None, query=None):
```
Comment (rayluo) [inline msal_extensions/_windows.py]: This time though, we would want to use **kwargs to accomodate possible future extra values introduced in the base class.

```python
def find(self, credential_type, **kwargs):
    ...
    return super(...).find(credential_type, **kwargs)
```

File: msal_extensions/_windows.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        if self._has_state_changed:
```
Comment (rayluo) [inline msal_extensions/_windows.py]: You mean `self._has_state_changed`**()**?

File: msal_extensions/_windows.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    """
+    _fields_ = [("cbData", wintypes.DWORD), ("pbData", ctypes.POINTER(ctypes.c_char))]
+
+    def raw(self):
+        # type: () -> bytes
+        cb_data = int(self.cbData)
+        pb_data = self.pbData
+        buffer = ctypes.c_buffer(cb_data)
+        _memcpy(buffer, pb_data, cb_data)
+        _local_free(pb_data)
```
[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: So, the local_free is definitely necessary as we don't want to create a memory leak! However, you're totally right that it should be safe to call this method twice in a way it just isn't right now. I'll fix that up.

File: msal_extensions/_windows.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            None,
+            None,
+            _CRYPTPROTECT_UI_FORBIDDEN,
+            ctypes.byref(result)
+        ):
+            return result.raw().decode('utf-8')
+        return u''
+
+
+class _WindowsTokenCache(msal.SerializableTokenCache):
```
[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 91c6badd2

File: msal_extensions/_windows.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in b9becbcd8

[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: So, I'm going to have to walk this change back. I got to a Windows machine and ran the tests, and curiously pulling this object out into the destructor causes a seg-fault that crashes the host Python process with it.

File: msal_extensions/_windows.py (hunk 3 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 9ef612f

File: msal_extensions/_windows.py (hunk 7 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 9ef612f

File: msal_extensions/_windows.py (hunk 4 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Hey, this is great news! If you don't mind, I'd like to keep this PR scoped to just getting this Windows agent targeting msal~=0.3.0. I can submit another PR to adopt 0.4.0.

File: msal_extensions/_windows.py (hunk 5 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Same as here: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/14#discussion_r286144472

File: msal_extensions/_windows.py (hunk 2 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 9ef612f

File: msal_extensions/_windows.py (hunk 1 above)
Comment (marstr) [reply msal_extensions/_windows.py]: I went ahead and renamed this method to `needs_refresh`, what do you think of that name?

File: msal_extensions/_windows.py (hunk 6 above)
Comment (marstr) [reply msal_extensions/_windows.py]: I believe this is feedback I should react to at the same time that I respond to this thread: https://github.com/AzureAD/microsoft-authentication-extensions-for-python/pull/14#discussion_r286144472

File: msal_extensions/_windows.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        buffer = ctypes.c_buffer(cb_data)
+        _memcpy(buffer, pb_data, cb_data)
+        _local_free(pb_data)
+        return buffer.raw
+
+
+# This code is modeled from a StackOverflow question, which can be found here:
+# https://stackoverflow.com/questions/463832/using-dpapi-with-python
+class WindowsDataProtectionAgent(object):
+    def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 082684b

Thanks for this by the way, I like the refactored version a lot better.

File: msal_extensions/_windows.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+class _WindowsTokenCache(msal.SerializableTokenCache):
+    DEFAULT_CACHE_LOCATION = os.path.join(os.getenv('LOCALAPPDATA'), '.IdentityService', 'msal.cache')
+    DEFAULT_ENTROPY = ''
+
+    def __init__(self, **kwargs):
+        super(_WindowsTokenCache, self).__init__()
+
+        self._cache_location = _WindowsTokenCache.DEFAULT_CACHE_LOCATION  # type: str
+        if 'cache_location' in kwargs:
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
```
[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in a712a73

File: msal_extensions/_windows.py (hunk 4 above)
Comment (rayluo) [reply msal_extensions/_windows.py]: Sure. This conversation was mean to be a FYI and not actionable until MSAL Python 0.4.0 is released.

Yet another interesting FYI: You might already know, that `msal~=0.3.0` notation means [automatically adopt every future versions with same major version number](https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires) rather than "0.3.0 only". It is the desirable notation and we use it all the time. Just keep in mind that this time we would want to manually patch it asap, AFTER msal python 0.4.0. :-)

I'll close this comment soon.

File: msal_extensions/_windows.py (hunk 6 above)
Comment (rayluo) [reply msal_extensions/_windows.py]: This one is about such `find(...)` itself, so it has nothing to do with the upcoming new `modify(...)`. It is just a future-proof notation to make this `find(...)` forward-compatible with any new keyworded parameter(s) introduced into base class's `find(...)` in the future. I would still suggest we do it right this time, once and for all. :-)

File: msal_extensions/_windows.py (hunk 4 above)
Comment (marstr) [reply msal_extensions/_windows.py]: > Yet another interesting FYI: You might already know, that msal~=0.3.0 notation means automatically adopt every future versions with same major version number rather than "0.3.0 only". It is the desirable notation and we use it all the time. Just keep in mind that this time we would want to manually patch it asap, AFTER msal python 0.4.0. :-)

I just tested it out by creating a new venv and running `pip install tropo~=0.1.0`, I confirmed that the behavior seems to be special-cased for the '0.x' major version band to treat minor versions as breaking. i.e installing tropo~=0.1.0 resolves to 0.1.4, not 0.2.1 even though [it's available](https://pypi.org/project/tropo/#history).

File: msal_extensions/_windows.py (hunk 1 above)
Comment (rayluo) [reply msal_extensions/_windows.py]: You ask for it! :-)

Per your request, I think the new name `needs_refresh` has the same mistake I made when I came up with the `has_state_changed` in the base class. They are both arguably vague in terms of direction. Should it be a need to handle 2-way sync in the future, the current name would become insufficient. That being said, the point is not to predict the future, because you can't. So, don't bother. :-)  Just one minor thing though. If my aforementioned mistake teaches me anything, it is that this `needs_refresh` behavior should largely be internal. So if you add a `_` prefix to make it `_needs_refresh`, it would be perfect because you can rename it later easily.

PS: Naming inside this WindowsTokenCache class matters, because it would become one of the major API of this extension module, even though we might have a factory for it.

Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in e4132a1

File: msal_extensions/_windows.py (hunk 6 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Fixed in 61104a9

File: msal_extensions/_windows.py (hunk 4 above)
Comment (marstr) [reply msal_extensions/_windows.py]: Oh, maybe we're talking about different things! Sorry, I read that as "As it is, you'll accidentally be pulling in 0.4.0", what you probably meant was "In order to fix this in the next PR, you'll need to remember to bump this version number."

Comment (rayluo) [reply msal_extensions/_windows.py]: Thanks for your experiment! I actually did not know that `~=` notation is "special-cased for the '0.x' major version band to treat minor versions as breaking. i.e installing tropo~=0.1.0 resolves to 0.1.4, not 0.2.1 even though it's available". That is a surprising "feature" to me.

Previously I tend to use the more explicit notation such as ">=0.1, <1" whose behaviour matches my expectation and mathematics: gives a highest available version within that range.

File: msal_extensions/_windows.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rayluo) [reply msal_extensions/_windows.py]: If that is the case, can we have one more try as to move that `_local_free()` out of `raw()`, and put it into `protect()` in this manner? On paper, it should work.

```python
    def protect(self, ...):
        ...
        try:
            return result.raw()
        finally:
            _local_free(result.pbData)
```

[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: Love it :) I'll give it another whack in the morning.

[the remark this answers is not in the document]
Comment (marstr) [reply msal_extensions/_windows.py]: So the trick ended up being that LocalFree **MUST NOT** be called memory allocated by ctypes.create_string_buffer, but MUST be called on memory allocated by the CryptProtectData and CryptUnprotectData functions. It just so happened that I only ever called `raw` on the latter, and never called `raw` for the former. This means that using `try` & `finally` allows us to manage the lifetime of only specific structs, where as moving it into a `__del__` method was breaking the contract.

Long story short, this is fixed in  161e6cf. On the otherhand, a4efa654 can be ignored.

[the remark this answers is not in the document]
Comment (rayluo) [reply msal_extensions/_windows.py]: Thanks for doing all the hard work to validate our latest fix, and documeting your finding! For the record, even if that 80e735c would work, I would still prefer the 2a25b5f approach. Because that's fundamentally the right pattern to manage memory (came from my 7 years on C/C++).

🚢
